File: local/receive_side.py
"""GAP #3: does FastAPI (and raw Starlette/ASGI) route QUERY AND read its body?"""
import subprocess
import sys
import threading
import time
import urllib.request

APP = '''
from fastapi import FastAPI, Request
app = FastAPI()

@app.api_route("/q", methods=["QUERY", "POST"])
async def q(request: Request):
    body = await request.body()
    return {"method": request.method, "body_len": len(body), "body": body.decode()}
'''
with open("recv_app.py", "w") as f:
    f.write(APP)

proc = subprocess.Popen([sys.executable, "-m", "uvicorn", "recv_app:app", "--port", "8692", "--log-level", "critical"])
time.sleep(4)

BODY = b'{"filter":"x","n":42}'
req = urllib.request.Request("http://127.0.0.1:8692/q", data=BODY, method="QUERY",
                            headers={"content-type": "application/json"})
try:
    with urllib.request.urlopen(req) as r:
        import json
        res = json.load(r)
        print("FastAPI ->", res)
        print("BODY_READ:", "YES" if res["body_len"] == len(BODY) else f"NO ({res['body_len']})",
              "| ROUTED:", "YES" if res["method"] == "QUERY" else "NO")
except Exception as e:  # noqa
    print("FastAPI FAIL:", e)
finally:
    proc.terminate()
    proc.wait()

File: local/receive_side.mjs
// GAP #3 (uncovered by the research): do server frameworks ROUTE a QUERY request
// AND READ its body? Many stacks body-gate on POST/PUT. Test Express + Fastify.
import express from 'express';
import Fastify from 'fastify';

const BODY = '{"filter":"x","n":42}';
const out = {};

// ---- Express ----
const ex = express();
ex.use(express.text({ type: '*/*' }));
ex.all('/q', (req, res) => res.json({ method: req.method, bodyLen: (req.body || '').length, body: req.body }));
const exSrv = await new Promise((ok) => { const s = ex.listen(8690, '127.0.0.1', () => ok(s)); });

// ---- Fastify (needs addHttpMethod for QUERY) ----
const fa = Fastify();
try { fa.addHttpMethod('QUERY', { hasBody: true }); } catch (e) { out.fastify_addmethod = 'FAILED: ' + e.message; }
fa.addContentTypeParser('*', { parseAs: 'string' }, (req, body, done) => done(null, body));
try {
  fa.route({ method: 'QUERY', url: '/q', handler: async (req) => ({ method: req.method, bodyLen: (req.body || '').length, body: req.body }) });
} catch (e) { out.fastify_route = 'FAILED: ' + e.message; }
await fa.listen({ port: 8691, host: '127.0.0.1' });

async function hit(port) {
  try {
    const r = await fetch(`http://127.0.0.1:${port}/q`, { method: 'QUERY', body: BODY, headers: { 'content-type': 'application/json' } });
    return { status: r.status, ...(await r.json().catch(() => ({}))) };
  } catch (e) { return { error: e.message }; }
}

out.express = await hit(8690);
out.fastify = await hit(8691);

for (const [k, v] of Object.entries(out)) console.log(k, '->', JSON.stringify(v));
console.log('\nBODY_READ:',
  'express', out.express?.bodyLen === BODY.length ? 'YES' : `NO (${out.express?.bodyLen})`,
  '| fastify', out.fastify?.bodyLen === BODY.length ? 'YES' : `NO (${out.fastify?.bodyLen})`);
exSrv.close(); await fa.close(); process.exit(0);
